pkg/validator/cluster.go

```diff
@@ -461,12 +461,30 @@ func validateAppSpecChanges(
 	for _, appRole := range targetApp.Spec.NodeRoles {
 		// First, ignore role image tags
 		// Second, ignore app config package url
+		roleImageChanged := false
 		prevAppRoleIdx := 0
-		for i, prevAppRole := range prevApp.Spec.NodeRoles {
+		for i := 0; i < len(prevApp.Spec.NodeRoles); i++ {
+			prevAppRole := &prevApp.Spec.NodeRoles[i]
 			if appRole.ID == prevAppRole.ID {
 				prevAppRoleIdx = i
-				prevApp.Spec.NodeRoles[prevAppRoleIdx].ImageRepoTag = appRole.ImageRepoTag
-				(&prevApp.Spec.NodeRoles[prevAppRoleIdx].SetupPackage.Info).PackageURL = appRole.SetupPackage.Info.PackageURL
+				if *prevAppRole.ImageRepoTag != *appRole.ImageRepoTag {
+					prevAppRole.ImageRepoTag = appRole.ImageRepoTag
+					roleImageChanged = true
+				}
+
+				if prevAppRole.SetupPackage.Info.PackageURL != appRole.SetupPackage.Info.PackageURL {
+					// https://github.com/bluek8s/kubedirector/issues/618
+					// If the role image was not changed but some other role attributes were changed
+					// decline the upgrade
```

Comment:
Took me a bit to get on board with this, but yeah I think this is correct.

Let me talk through how I see it.

Recap of the requirement: if the kdapp spec for a kdcluster was changed -- which must be the case since we're in this function -- then we want the current containers in a given role to either be completely unaffected by the switch between the new and old kdapp, or the role must change its image.

We currently only allow two changes in the kdapp: package URL and cardinality. Cardinality doesn't affect the current containers. If package URL is changing, we make sure the image is changing too.

Seems OK!